code/modules/events/crystal_event.dm

```diff
@@ -0,0 +1,412 @@
+	startWhen = 10
+	announceWhen = 1
+	///Is the name of the wave, used to check wich wave will be generated
+	var/wave_name
+	///Max number of portals that can spawn per type of wave
+	var/portal_numbers
+
+/datum/round_event/crystal_invasion/start()
+	choose_wave_type()
+
+/datum/round_event/crystal_invasion/announce(fake)
+	priority_announce("WARNING - Destabilization of the Supermatter Crystal Matrix detected, please stand by waiting further instructions", "Alert")
+	sound_to_playing_players('sound/misc/notice1.ogg')
+
+///Choose the type of the wave
+/datum/round_event/crystal_invasion/proc/choose_wave_type()
+	if(!wave_name)
+		wave_name = pickweight(list(
+			"small wave" = 45,
+			"medium wave" = 35,
+			"big wave" = 15,
+			"huge wave" = 5))
+	switch(wave_name)
+		if("small wave")
+			portal_numbers = pick(2, 3, 4, 5)
+		if("medium wave")
+			portal_numbers = pick(4, 5, 6, 7)
+		if("big wave")
+			portal_numbers = pick(6, 7, 8, 9)
+		if("huge wave")
+			portal_numbers = pick(9, 10, 11, 12)
+		else
+			WARNING("Wave name of [wave_name] not recognised.")
+			kill()
+
+	var/list/sm_crystal = list()
+	for(var/obj/machinery/power/supermatter_crystal/temp in GLOB.machines)
+		if(QDELETED(temp))
+			WARNING("No engine found.")
+			kill()
+		if(istype(temp, /obj/machinery/power/supermatter_crystal/shard))
+			message_admins("found a shard, skipping it")
+			continue
+		sm_crystal += temp
+	if(sm_crystal == null)
+		WARNING("No engine found.")
+		kill()
+	var/obj/machinery/power/supermatter_crystal/crystal = pick(sm_crystal)
+	crystal.destabilize(portal_numbers)
+
+	priority_announce("WARNING - Numerous energy fluctuations have been detected from your Supermatter; we estimate a [wave_name] of crystalline creatures \
+						coming from \[REDACTED]; there will be [portal_numbers] portals spread around the station that you must close. Harvest a \[REDACTED] \
+						anomaly from a portal, place it inside a crystal stabilizer, and inject it into your Supermatter to stop a ZK-Lambda-Class Cosmic Fragmentation Scenario from occurring.", "Alert", 'sound/misc/notice1.ogg')
+
+	addtimer(CALLBACK(src, .proc/spawn_portals), 10 SECONDS)
+
+/datum/round_event/crystal_invasion/proc/spawn_portals()
+	for(var/i = 0, i< portal_numbers, i++)
+		spawn_portal(GLOB.waves[wave_name])
+
+///Spawn one portal in a random location choosen from the generic_event_spawns list
+/datum/round_event/crystal_invasion/proc/spawn_portal(list/wave_type)
+	var/list/spawners = list()
+	for(var/obj/effect/landmark/event_spawn/temp in GLOB.generic_event_spawns)
+		if(QDELETED(temp))
+			continue
+		if(is_station_level(temp.loc.z))
+			spawners += temp
+
+	if(!spawners.len)
+		message_admins("No landmarks on the station, aborting")
+		return MAP_ERROR
+
+	var/pick_portal = pickweight(wave_type)
+	var/obj/spawner = pick_n_take(spawners)
+	new pick_portal(spawner.loc)
+
+/*
+This section is for the destabilized SM
+*/
+
+/obj/machinery/destabilized_crystal
```

Comment:
why is this its own separate object instead of a variable or something for the SM crystal?

Reply:
because is better to have another object to handle this instead of the SM to make it more modular (we can change this obj far more easily) and to not bloat too much the SM code